org.mixedrealitytoolkit.input/InteractionModes/NearInteractionModeDetector.cs

```diff
@@ -16,58 +16,101 @@ namespace MixedReality.Toolkit.Input
     [AddComponentMenu("MRTK/Input/Near Interaction Mode Detector")]
     public class NearInteractionModeDetector : ProximityDetector
     {
+        /// <summary>
+        /// The set of near interactors that belongs to near interaction
+        /// </summary>
         [SerializeField]
         [Tooltip("The set of near interactors that belongs to near interaction")]
         private List<XRBaseInteractor> nearInteractors;
 
         /// <summary>
-        /// Used to keep track of the previously detected colliders so that we can know which
-        /// colliders stopped being detected and update their buttons front plate and rounded
-        /// rect if they are labeled as dynamic (based on proximity).
+        /// Keeps track of the previously detected interactables so that we can know which
+        /// interactable stopped being detected and trigger corresponding event.
+        /// </summary>
+        private readonly HashSet<IXRProximityInteractable> previouslyDetectedInteractables = new();
+
+        /// <summary>
+        /// Stores the currently detected interactables (updated every frame).
         /// </summary>
-        List<Collider> previouslyDetectedColliders = new List<Collider>();
+        private readonly List<IXRProximityInteractable> currentlyDetectedInteractables = new();
+
+        /// <summary>
+        /// Stores the interactables that are no longer detected (updated every frame).
+        /// </summary>
+        private readonly List<IXRProximityInteractable> noLongerDetectedInteractables = new();
 
         /// <inheritdoc />
         public override bool IsModeDetected()
         {
             bool result = base.IsModeDetected() || IsNearInteractorSelecting();
             if (result)
             {
-                for (int i = 0; i < previouslyDetectedColliders.Count; i++)
+                UpdateCurrentlyDetectedInteractables();
+                UpdateProximityExited();
+                UpdateProximityEntered();
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Calls OnProximityExited for all interactables that were previously ProximityHover detected but are no longer detected.
+        /// </summary>
+        private void UpdateProximityExited()
+        {
+            noLongerDetectedInteractables.Clear();
+
+            foreach (IXRProximityInteractable previouslyDetectedInteractable in previouslyDetectedInteractables)
+            {
+                if (!currentlyDetectedInteractables.Contains(previouslyDetectedInteractable))
+                {
+                    noLongerDetectedInteractables.Add(previouslyDetectedInteractable);
+                }
+            }
+
+            foreach (IXRProximityInteractable noLongerDetectedInteractable in noLongerDetectedInteractables)
+            {
+                noLongerDetectedInteractable.OnProximityExited(new ProximityHoverExitedEventArgs(this));
```

Comment:
Does this component protect against previous objects having been destroyed?  Is it possible that the noLongerDetectedInteractable is destroyed at this time, so calling OnProximityExited() will cause a null ref?

Perhaps we should check for Unity's overridden object null check before proceeding?
`if (noLongerDetectedInteractable != null) ...`

Reply:
Add null check in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/659/commits/c97fcc6a878f218626f5099c763284ad6118699d, resolving.